File: python/example/samples/flag.py
"""
File: flag.py
Project: samples
Created Date: 14/09/2023
Author: Shun Suzuki
-----
Last Modified: 14/09/2023
Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
-----
Copyright (c) 2023 Shun Suzuki. All rights reserved.

"""


import threading

from pyautd3 import ConfigureForceFan, ConfigureReadsFPGAInfo, Controller


async def flag(autd: Controller) -> None:
    print("press any key to run fan...")
    _ = input()

    await autd.send_async(ConfigureReadsFPGAInfo(lambda _: True), ConfigureForceFan(lambda _: True))

    fin = False

    def f() -> None:
        nonlocal fin
        print("press any key stop checking FPGA status...")
        _ = input()
        fin = True

    th = threading.Thread(target=f)
    th.start()

    prompts = ["-", "/", "|", "\\"]
    prompts_idx = 0
    while not fin:
        states = await autd.fpga_info_async()
        print(f"{prompts[(prompts_idx // 1000) % len(prompts)]} FPGA Status...")
        print("\n".join([f"\x1b[0K[{i}]: thermo = {state.is_thermal_assert()}" for i, state in enumerate(states)]))
        print(f"\x1b[{len(states) + 1}A", end="")
        prompts_idx += 1
    print("\x1b[1F\x1b[0J")

    th.join()
    await autd.send_async(ConfigureReadsFPGAInfo(lambda _: False), ConfigureForceFan(lambda _: False))

File: dotnet/cs/example/Samples/Flag.cs
/*
 * File: Flag.cs
 * Project: Samples
 * Created Date: 13/09/2023
 * Author: Shun Suzuki
 * -----
 * Last Modified: 08/12/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2023 Shun Suzuki. All rights reserved.
 * 
 */

using AUTD3Sharp;

namespace Samples;

internal static class FlagTest
{
    public static async Task Test<T>(Controller<T> autd)
    {
        Console.WriteLine("press any key to run fan...");
        Console.ReadKey(true);

        await autd.SendAsync(new ConfigureForceFan(_ => true), new ConfigureReadsFPGAInfo(_ => true));

        var fin = false;
        var th = Task.Run(() =>
        {
            Console.WriteLine("press any key stop checking FPGA status...");
            Console.ReadKey(true);
            fin = true;
        });

        var prompts = new[] { '-', '/', '|', '\\' };
        var promptsIdx = 0;
        while (!fin)
        {
            var states = await autd.FPGAInfoAsync();
            Console.WriteLine($"{prompts[promptsIdx++ / 1000 % prompts.Length]} FPGA Status...");
            Console.WriteLine(string.Join("\n", states));
            Console.Write($"\x1b[{states.Length + 1}A");
        }

        th.Wait();

        await autd.SendAsync(new ConfigureForceFan(_ => false), new ConfigureReadsFPGAInfo(_ => false));
    }
}
